Add Unity and CMock to NCS

nordic-krch opened the pull request:
This PR contains 3 commits:
- adding cmock repository to ncs (extending west configuration).
- adding files (boilerplate, cmakelists, kconfig, etc) which enables adding unity tests.
- adding example (tests/unity/example_test) to show how unit tests can be created.

Open items:
- currently cmock repository is taken directly from https://github.com/ThrowTheSwitch. Should we fork it to nordic playground and take it from there?
- currently cmock and unity CMakeLists.txt is created in nrf/tests/unity and it points to the sources in external repository (this way repository is unspoiled). Any better proposal?

Please note that this is my first PR to NCS. Let me know if i did something wrong.
Please note that i'm total newbie to cmake so be patient in case of rookie mistakes.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/NordicPlayground/fw-nrfconnect-nrf?pullRequest=519) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/NordicPlayground/fw-nrfconnect-nrf?pullRequest=519) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/NordicPlayground/fw-nrfconnect-nrf?pullRequest=519) it.</sub>

NordicBuilder commented:
Found the following issues, please fix and resubmit:

## Checkpatch issues

```
No additional types will be considered - file 'scripts/checkpatch/typedefsfile': No such file or directory
-:147: ERROR:TRAILING_WHITESPACE: trailing whitespace
#147: FILE: scripts/unity/func_name_list.py:38:
+       $

-:183: ERROR:TRAILING_WHITESPACE: trailing whitespace
#183: FILE: scripts/unity/header_prepare.py:29:
+    $

-:185: ERROR:TRAILING_WHITESPACE: trailing whitespace
#185: FILE: scripts/unity/header_prepare.py:31:
+    content = syscall_decl_pattern.sub("", content) $

-:189: ERROR:TRAILING_WHITESPACE: trailing whitespace
#189: FILE: scripts/unity/header_prepare.py:35:
+    $

-:195: ERROR:TRAILING_WHITESPACE: trailing whitespace
#195: FILE: scripts/unity/header_prepare.py:41:
+    f_wrap.close() $

-:205: ERROR:TRAILING_WHITESPACE: trailing whitespace
#205: FILE: scripts/unity/header_prepare.py:51:
+       $

-:220: ERROR:TRAILING_WHITESPACE: trailing whitespace
#220: FILE: tests/unity/CMakeLists.txt:8:
+        ${ZEPHYR_BASE}/../cmock/vendor/unity/src $

-:497: ERROR:TRAILING_WHITESPACE: trailing whitespace
#497: FILE: tests/unity/unity_test.cmake:27:
+    COMMAND ${ruby_cmd} ${cmock_script} $

-:498: ERROR:TRAILING_WHITESPACE: trailing whitespace
#498: FILE: tests/unity/unity_test.cmake:28:
+    ${cmock_prefix} ${cmock_plugins} $

-:513: ERROR:TRAILING_WHITESPACE: trailing whitespace
#513: FILE: tests/unity/unity_test.cmake:43:
+    $

-:566: ERROR:TRAILING_WHITESPACE: trailing whitespace
#566: FILE: tests/unity/unity_test.cmake:96:
+    $

-:577: ERROR:TRAILING_WHITESPACE: trailing whitespace
#577: FILE: tests/unity/unity_test.cmake:107:
+                                                                               $

-:578: ERROR:TRAILING_WHITESPACE: trailing whitespace
#578: FILE: tests/unity/unity_test.cmake:108:
+    execute_process(                                                            $

-:579: ERROR:TRAILING_WHITESPACE: trailing whitespace
#579: FILE: tests/unity/unity_test.cmake:109:
+      COMMAND                                                                   $

-:580: ERROR:TRAILING_WHITESPACE: trailing whitespace
#580: FILE: tests/unity/unity_test.cmake:110:
+      "python"                                                                  $

-:581: ERROR:TRAILING_WHITESPACE: trailing whitespace
#581: FILE: tests/unity/unity_test.cmake:111:
+      ${HEADER_PREPARE_SCRIPT}                                                       $

-:582: ERROR:TRAILING_WHITESPACE: trailing whitespace
#582: FILE: tests/unity/unity_test.cmake:112:
+      "--input" ${in_header} $

-:584: ERROR:TRAILING_WHITESPACE: trailing whitespace
#584: FILE: tests/unity/unity_test.cmake:114:
+      "--wrap" ${wrap_header}                                            $

-:585: ERROR:TRAILING_WHITESPACE: trailing whitespace
#585: FILE: tests/unity/unity_test.cmake:115:
+      WORKING_DIRECTORY ${PROJECT_SOURCE_DIR}                                   $

-:586: ERROR:TRAILING_WHITESPACE: trailing whitespace
#586: FILE: tests/unity/unity_test.cmake:116:
+      RESULT_VARIABLE op_result                                                 $

-:587: ERROR:TRAILING_WHITESPACE: trailing whitespace
#587: FILE: tests/unity/unity_test.cmake:117:
+      OUTPUT_VARIABLE output_result                                             $

-:588: ERROR:TRAILING_WHITESPACE: trailing whitespace
#588: FILE: tests/unity/unity_test.cmake:118:
+    )                         $

-:597: ERROR:TRAILING_WHITESPACE: trailing whitespace
#597: FILE: tests/unity/unity_test.cmake:127:
+    $

- total: 23 errors, 0 warnings, 519 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

NOTE: Whitespace errors detected.
      You may wish to use scripts/cleanpatch or scripts/cleanfile

Your patch has style problems, please review.

NOTE: Ignored message types: AVOID_EXTERNS BRACES CONFIG_EXPERIMENTAL CONST_STRUCT DATE_TIME FILE_PATH_CHANGES MINMAX NETWORKING_BLOCK_COMMENT_STYLE PRINTK_WITHOUT_KERN_LEVEL SPLIT_STRING VOLATILE

NOTE: If any of the errors are false positives, please report
      them to the maintainers.

```

lemrey commented:
@nordic-krch fantastic! Are there any requirements in terms of tooling for this, e.g. Ruby, Python modules?

nordic-krch commented:
@lemrey, right. Ruby is of course required. In python no additional lib is needed.

nordic-krch commented:
@SebastianBoe @tejlmand could you take a look as this is mainly cmake stuff and i'm newbie there.
fyi @gbakke

SebastianBoe commented:
regex parsing of C?

:(

I'm concerned that we won't be able to regex parse the full language of CPP+CC, and that what we have here is just able to parse a subset of the language.

What happens when the parsing fails, can we create mocks manually for the cases that are not covered?

Sorry if this question doesn't make sense, I'm not familiar with mocking frameworks.

SebastianBoe commented on scripts/unity/func_name_list.py:
```diff
@@ -0,0 +1,39 @@
+import re
+import argparse
+
+def func_names_from_header(in_file, out_file):
```
Do you just need the func_names? Not the full prototype?

If so, maybe it would be possible to compile the header and then take the func names from the symbol table.

Then we don't need to regex parse unpreprocessed C code.

SebastianBoe commented on scripts/unity/func_name_list.py:
```diff
@@ -0,0 +1,39 @@
+    content = f_in.read()
+    f_in.close()
+
+    f_out = open(out_file,'w')
```
files must be closed, e.g. by using 'with'.

nordic-krch commented:
@SebastianBoe it is parsing header files mainly in search for function declarations and static inline functions. We need to handle following:
- static inline functions cannot be mocked (because implementation is already provided in the header)
- zephyr specific function declaration in case of syscall
- wrapping around calls to allow mocking without removing mocked functions from the binary.

So in regexp we do not go beyond function declarations and we are focused on C since Cmock/Unity is C unit test framework.

SebastianBoe commented on scripts/unity/header_prepare.py:
```diff
@@ -0,0 +1,51 @@
+    static_inline_pattern = re.compile(r"((__deprecated )?)(static inline )((struct )?)(\w+ )(\*?)((_impl_)?)(\w+\([^\)]+\))(\n\{[\s\S]+?\n\})", re.M | re.S)
+    (content, static_inline_cnt) = static_inline_pattern.subn(r"\4\6\7\10;", content)
+
+    print static_inline_cnt
```
print() is best-practice

SebastianBoe commented:
> we are focused on C 

Sorry, by CPP, I meant C preprocessor.

nordic-krch commented:
@SebastianBoe i've applied your comments. Regarding compiling header file: I'm not sure how to do that but imo, it can be later on improved if we find issues with using python to parse the header.

> What happens when the parsing fails, can we create mocks manually for the cases that are not covered?

In case regexp is missing something and we find case which fails we need to improve regexp.

SebastianBoe commented:
> In case regexp is missing something and we find case which fails we need to improve regexp.

What if it's impossible?

e.g. if the prototype is constructed by the preprocessor?

nordic-krch commented:
@SebastianBoe, worst case function won't be mocked. Header processed by python next goes to cmock ruby script which processes it in search for function prototypes. It will create mocks only for found prototypes. Python script can only improve.

SebastianBoe commented:
Okay, conceptually this makes sense then.

I can review the CMake code when the PR is in PR-form and no longer an RFC.

nordic-krch commented:
i've put RFC since there were some open items mentioned in the PR but i see no opinions so far and those open items can be fixed later (like changing west to use unity fork instead original).

SebastianBoe commented:
@gbakke , @hakonfam : FYI

SebastianBoe commented:
> currently cmock repository is taken directly from https://github.com/ThrowTheSwitch. Should we fork it to nordic playground and take it from there?

I'd postpone this until necessary.

SebastianBoe commented on CMakeLists.txt:
```diff
@@ -13,3 +13,4 @@ add_subdirectory(ext)
 add_subdirectory(lib)
 add_subdirectory(subsys)
 add_subdirectory(drivers)
+add_subdirectory(tests/unity)
```
I've seen people be lazy and do add_subdirectory(tests/junit) instead of re-organizing this into add_subdirectory(tests), so let's do the future-proof solution from the get go.

Re-write this to not add_subdirectory several levels down.

hakonfam commented on tests/unity/example_test/src/example_test.h:
```diff
@@ -0,0 +1,26 @@
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+#ifndef __EXAMPLE_TEST_H
```
Remove `__` prefix in all the header guards, AFAICS the style used is like `AT_CMD_PARSER_H_`

lemrey replied:
or the longer `ZEPHYR_INCLUDE_`

SebastianBoe commented on scripts/unity/func_name_list.py:
```diff
@@ -0,0 +1,35 @@
+    elif args.postfix:
+        out_name = args.input + args.postfix
+    else:
+        out_name = args.input + ".flist"
```
This case is not needed and can be dropped, it is safe and clean to limit support to only args.output.

SebastianBoe commented on scripts/unity/func_name_list.py:
```diff
@@ -0,0 +1,35 @@
+    parser = argparse.ArgumentParser()
+    parser.add_argument("-i", "--input", type=str, help="input header file", required=True)
+    parser.add_argument("-o", "--output", type=str, help="output function list file")
+    parser.add_argument("-p", "--postfix", type=str, help="output file as input file with postfix")
```
postfix is not in use and should be dropped.

Just supporting --output is simple and works fine.

SebastianBoe commented on scripts/unity/func_name_list.py:
```diff
@@ -0,0 +1,35 @@
+    with open(in_file) as f_in:
+        content = f_in.read()
+
+        with open(out_file,'w') as f_out:
```
You can unindent this, 'content' does not go out of scope.

SebastianBoe commented on scripts/unity/func_name_list.py:
```diff
@@ -0,0 +1,35 @@
+        content = f_in.read()
+
+        with open(out_file,'w') as f_out:
+            x= re.findall(r"(struct\s|\s?)(\w+\s)(\*?)(\w+?)(\(.*?\))(;)", content, re.M | re.S)
```
It is implied from the function name. But a comment like:

" Regex match all function names in the header file"

would be helpful

SebastianBoe commented on scripts/unity/header_prepare.py:
```diff
@@ -0,0 +1,47 @@
+    with open(in_file) as f_in:
+        content = f_in.read()
+
+    if content == None:
```
AFAIK read() does not return None

SebastianBoe commented on scripts/unity/header_prepare.py:
```diff
@@ -0,0 +1,47 @@
+import re
+import argparse
+
+def func_names_from_header(in_file, out_file, out_wrap_file):
```
Should use a different function name since it clearly does something else than the other func_names_from_header

SebastianBoe commented on scripts/unity/header_prepare.py:
```diff
@@ -0,0 +1,47 @@
+
+    # change static inline functions to normal function declaration
+    static_inline_pattern = re.compile(r"((__deprecated )?)(static inline )((struct )?)(\w+ )(\*?)((_impl_)?)(\w+\([^\)]+\))(\n\{[\s\S]+?\n\})", re.M | re.S)
+    (content, static_inline_cnt) = static_inline_pattern.subn(r"\4\6\7\10;", content)
```
What about these macro's?

INLINE
__STATIC_INLINE
ALWAYS_INLINE

nordic-krch replied:
Added __STATIC_INLINE and ALWAYS_INLINE, I haven't seen INLINE being used in API. Added also examples of static inline functions to the example test project.

SebastianBoe commented on scripts/unity/header_prepare.py:
```diff
@@ -0,0 +1,47 @@
+        return
+
+    # change static inline functions to normal function declaration
+    static_inline_pattern = re.compile(r"((__deprecated )?)(static inline )((struct )?)(\w+ )(\*?)((_impl_)?)(\w+\([^\)]+\))(\n\{[\s\S]+?\n\})", re.M | re.S)
```
Put a newline after compile(

to shorten the line with the regex.

SebastianBoe commented on tests/unity/CMakeLists.txt:
```diff
@@ -0,0 +1,13 @@
+#
+# SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+#
+
```
set(CMOCK_DIR ${ZEPHYR_BASE}/../cmock)

nordic-krch commented on tests/unity/unity_test.cmake:
```diff
@@ -0,0 +1,147 @@
+# File is then passed to 'cmock_linker_trick' which adds linker option for each
+# function listed in the file.
+function(cmock_linker_wrap_trick header_file_path)
+    set(FUNC_LIST_SCRIPT "$ENV{ZEPHYR_BASE}/../nrf/scripts/unity/func_name_list.py")
```
that didn't work for me. Isn't it because i used boilerplace.cmake from zephyr?

SebastianBoe replied:
Ah, this is not a part of the zephyr build system, is it?

Never mind then, and never mind about the ZEPHYR_BASE variable.

nordic-krch replied:
I would like to solve that one as now there is a fixed dependency on nrf repo location and that should come from `NRF_DIR`. However, i somehow don't have access to that variable in functions called by test CMakeLists.txt file. NRF_DIR is present in library CMakeLists.txt. How is it done for `ZEPHYR_BASE` which is visible everywhere?

nordic-krch replied:
@SebastianBoe do you idea why NRF_DIR is not visible there?

SebastianBoe replied:
NRF_DIR is de-referenced using the scope of the caller.

The caller is the 'app' directory.

NRF_DIR is set in the

app --> zephyr/CMakeListst.txt --> nrf/CMakeLists.txt

directory, and is not visible in zephyr/CMakeListst.txt or 'app'. Only in nrf/CMakeLists.txt and sub-directories of nrf/CMakeLists.txt. (subdirectories, as determined by the add_subdirectory hierarchy, not the filesystem hierarchy).

The standard solution here would be to use a global property. But this feels like overkill, so ../nrf is fine for now IMHO.

nordic-krch replied:
isn't `nrf` a suggestion and user could pick different folder?

SebastianBoe commented on tests/unity/unity_test.cmake:
```diff
@@ -0,0 +1,135 @@
+set(UNITY_PRODUCTS_DIR "${APPLICATION_BINARY_DIR}/unity")
+set(CMOCK_PRODUCTS_DIR "${UNITY_PRODUCTS_DIR}/mocks")
+set(CMOCK_DIR "${ZEPHYR_BASE}/../cmock")
+set(RUBY_EXEC "/usr/bin/ruby")
```
find_package(Ruby \<whatever the CMock minimimum supported Ruby version is>)

/usr/bin/ruby will then be found in ${RUBY_EXECUTABLE}.

https://github.com/Kitware/CMake/blob/master/Modules/FindRuby.cmake

nordic-krch replied:
I've tried to add that but it resulted in some errors which are confusing. Errors mixed with `found suitable version "2.5.1", minimum required is "2.0.0"`. Do you have any suggestions?
```
CMake Error: Error required internal CMake variable not set, cmake may not be built correctly.
Missing variable is:
CMAKE_FIND_LIBRARY_PREFIXES
CMake Error: Error required internal CMake variable not set, cmake may not be built correctly.
Missing variable is:
CMAKE_FIND_LIBRARY_SUFFIXES
-- Could NOT find Ruby (missing: RUBY_INCLUDE_DIR RUBY_CONFIG_INCLUDE_DIR) (found suitable version "2.5.1", minimum required is "2.0.0")
```

SebastianBoe replied:
I was not able to figure this out either.

Since we don't need the ruby include dirs yet, I'd suggest that we use the simpler find_program instead.

So

```
find_program(
  RUBY
  ruby
  )
```

SebastianBoe commented on tests/unity/unity_test.cmake:
```diff
@@ -0,0 +1,135 @@
+    --mock_path=${dst_path}
+    ${header_path}
+    WORKING_DIRECTORY ${PROJECT_SOURCE_DIR}
+    RESULT_VARIABLE op_result
```
Should we abort when op_result != 0 ?

nordic-krch replied:
added fatal_error on errors returned by commands.

SebastianBoe commented on tests/unity/unity_test.cmake:
```diff
@@ -0,0 +1,135 @@
+    ${header_path}
+    WORKING_DIRECTORY ${PROJECT_SOURCE_DIR}
+    RESULT_VARIABLE op_result
+    OUTPUT_VARIABLE output_result
```
When the command fails, is there anything in this variable that we would want to show the user?

nordic-krch replied:
added

SebastianBoe commented on tests/unity/unity_test.cmake:
```diff
@@ -0,0 +1,135 @@
+# Add --wrap linker option for each function listed in the input file.
+function(cmock_linker_trick func_name_path)
+    file(READ "${func_name_path}" contents)
+    string(REGEX REPLACE "\n" ";" contents "${contents}")
```
Use file(STRINGS, instead of file and string.

https://cmake.org/cmake/help/v3.14/command/file.html#strings

nordic-krch replied:
i cannot find how if it's possible to do file(STRINGS REGEX REPLACE..).

SebastianBoe commented on tests/unity/unity_test.cmake:
```diff
@@ -0,0 +1,135 @@
+#
```
Is it possible to avoid this file by organizing the CMock generation as a Zephyr library / subsystem ?

I have not seen anything so far that couldn't be moved out of this file at least ...

nordic-krch replied:
i don't know. im a cmake newbie. but wouldn't that require to move that upstream? I'm not sure if we want to keep it ncs or move to zephyr.

SebastianBoe replied:
No, it wouldn't require moving to upstream. nrf can define Zephyr libraries.

Generally, for new software, we want to organize it as libraries. And then have application's enable that software by including boilerplate.cmake, and then enabling it through Kconfig.

This PR uses a different boilerplate.cmake to express that a piece of SW should be enabled. This is unfortunate because it diverges from how we generally organize things.

Now, it could be that we for technical reasons need to do so, but before we do something different, we need to try to stick to the system.

nordic-krch replied:
I moved everything to CMAkeLists.txt and removed unity_test.cmake. I had one issue: when i'm setting variable in library it is not seen in functions. For time being i've added copy of those variables in functions. How to define variables that will have global scope?

SebastianBoe replied:
Discussed offline

nordic-krch commented on tests/Kconfig:
```diff
@@ -0,0 +1,7 @@
+# SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+#
+
+rsource "unity/Kconfig"
```
what about this PR: https://github.com/zephyrproject-rtos/zephyr/pull/14452 where new lines at the  end of file was removed?

SebastianBoe replied:
That PR is changing from

`line\n\nEOF`

to

`line\nEOF`

what we have here is

`lineEOF`

SebastianBoe replied:
This is still not resolved, is it? Is there a misunderstanding?

nordic-krch replied:
no, i fixed in others but somehow missed that one. fixed now.

SebastianBoe commented:
```
Zephyr version: 1.14.0
-- Found PythonInterp: /home/sebo/bin/python3.4 (found suitable version "3.5.2", minimum required is "3.4") 
-- Selected BOARD nrf52840_pca10056
-- Loading /home/sebo/ncs/zephyr/boards/arm/nrf52840_pca10056/nrf52840_pca10056.dts as base
-- Overlaying /home/sebo/ncs/zephyr/dts/common/common.dts
nrf52840_pca10056.dts_compiled: Warning (simple_bus_reg): /soc/virtualcom: missing or empty reg/ranges property
Parsing Kconfig tree in /home/sebo/ncs/nrf/tests/unity/example_test/Kconfig
Loading /home/sebo/ncs/zephyr/boards/arm/nrf52840_pca10056/nrf52840_pca10056_defconfig as base
Merging /home/sebo/ncs/nrf/tests/unity/example_test/prj.conf
Configuration written to '/home/sebo/ncs/nrf/tests/unity/example_test/b/zephyr/.config'
-- Cache files will be written to: /home/sebo/.cache/zephyr
-- The C compiler identification is GNU 8.2.1
-- The CXX compiler identification is GNU 8.2.1
-- The ASM compiler identification is GNU
-- Found assembler: /home/sebo/Downloads/gnu-mcu-eclipse/arm-none-eabi-gcc/8.2.1-1.4-20190214-0604/bin/arm-none-eabi-gcc
-- Performing Test toolchain_is_ok
-- Performing Test toolchain_is_ok - Success
Including module(s): nrf;nrfxlib
-- /usr/bin/ruby
/usr/bin/ruby: No such file or directory -- /home/sebo/ncs/zephyr/../cmock/vendor/unity/auto/generate_test_runner.rb (LoadError)
CMake Error at /home/sebo/ncs/nrf/tests/unity/CMakeLists.txt:53 (message):
  Failed to generate runner
  /home/sebo/ncs/nrf/tests/unity/example_test/b/runner/runner/runner_example_test.c
Call Stack (most recent call first):
  CMakeLists.txt:12 (test_runner_generate)


-- Configuring incomplete, errors occurred!
See also "/home/sebo/ncs/nrf/tests/unity/example_test/b/CMakeFiles/CMakeOutput.log".
See also "/home/sebo/ncs/nrf/tests/unity/example_test/b/CMakeFiles/CMakeError.log".
sebo@mach:~/ncs/nrf/tests/unity/example_test$ ls ~/ncs/cmock/vendor/unity/
sebo@mach:~/ncs/nrf/tests/unity/example_test$ 
```

Any idea what this error is? Is west not cloning submodules perhaps?

SebastianBoe commented:
I can confirm that it was a submodule issue.

SebastianBoe commented on tests/unity/example_test/CMakeLists.txt:
```diff
@@ -0,0 +1,34 @@
+#
```
I'd like a .yml file or a .rst file that explains which boards I can expect this to work on.

SebastianBoe replied:
Also, which boards should this work on?

native_posix?

all nrf boards?

SebastianBoe replied:
I'm getting a build failure with nrf52840_pca10056:

```
Zephyr version: 1.14.0
-- Found PythonInterp: /home/sebo/bin/python3.4 (found suitable version "3.5.2", minimum required is "3.4") 
-- Selected BOARD nrf52840_pca10056
-- Loading /home/sebo/ncs/zephyr/boards/arm/nrf52840_pca10056/nrf52840_pca10056.dts as base
-- Overlaying /home/sebo/ncs/zephyr/dts/common/common.dts
nrf52840_pca10056.dts_compiled: Warning (simple_bus_reg): /soc/virtualcom: missing or empty reg/ranges property
Parsing Kconfig tree in /home/sebo/ncs/nrf/tests/unity/example_test/Kconfig
Loading /home/sebo/ncs/zephyr/boards/arm/nrf52840_pca10056/nrf52840_pca10056_defconfig as base
Merging /home/sebo/ncs/nrf/tests/unity/example_test/prj.conf
Configuration written to '/home/sebo/ncs/nrf/tests/unity/example_test/b/zephyr/.config'
-- Cache files will be written to: /home/sebo/.cache/zephyr
-- The C compiler identification is GNU 8.2.1
-- The CXX compiler identification is GNU 8.2.1
-- The ASM compiler identification is GNU
-- Found assembler: /home/sebo/Downloads/gnu-mcu-eclipse/arm-none-eabi-gcc/8.2.1-1.4-20190214-0604/bin/arm-none-eabi-gcc
-- Performing Test toolchain_is_ok
-- Performing Test toolchain_is_ok - Success
Including module(s): nrf;nrfxlib
-- /usr/bin/ruby
-- RUBY_EXECUTABLE: /usr/bin/ruby
-- Generating test runner 
-- Generating cmock for header src/foo/foo.h
-- Configuring done
-- Generating done
-- Build files have been written to: /home/sebo/ncs/nrf/tests/unity/example_test/b
[1/117] Preparing syscall dependency handling

[2/117] Generating include/generated/syscall_macros.h
[3/117] Generating misc/generated/syscalls_subdirs.trigger
[4/117] Generating include/generated/kobj-types-enum.h, include/generated/otype-to-str.h
[5/117] Generating include/generated/driver-validation.h
[6/117] Generating misc/generated/syscalls.json
[7/117] Generating include/generated/syscall_dispatch.c, include/generated/syscall_list.h
[8/117] Building C object zephyr/CMakeFiles/offsets.dir/arch/arm/core/offsets/offsets.c.obj
[9/117] Generating include/generated/offsets.h
[10/117] Building C object CMakeFiles/app.dir/src/foo/foo.c.obj
[11/117] Building C object CMakeFiles/app.dir/src/uut/uut.c.obj
[12/117] Building C object zephyr/CMakeFiles/zephyr.dir/lib/os/crc16_sw.c.obj
[13/117] Building C object zephyr/CMakeFiles/zephyr.dir/lib/os/crc32_sw.c.obj
[14/117] Building C object zephyr/CMakeFiles/zephyr.dir/lib/os/crc8_sw.c.obj
[15/117] Building C object zephyr/CMakeFiles/zephyr.dir/lib/os/crc7_sw.c.obj
[16/117] Building C object CMakeFiles/app.dir/runner/runner/runner_example_test.c.obj
FAILED: CMakeFiles/app.dir/runner/runner/runner_example_test.c.obj 
ccache /home/sebo/Downloads/gnu-mcu-eclipse/arm-none-eabi-gcc/8.2.1-1.4-20190214-0604/bin/arm-none-eabi-gcc -DBUILD_VERSION=v1.14.0-rc1-896-g578dac80f7c3 -DKERNEL -DNRF52840_XXAA -D_FORTIFY_SOURCE=2 -D__ZEPHYR__=1 -Imocks -I../src -I.././src/uut -I../. -I/home/sebo/ncs/zephyr/kernel/include -I/home/sebo/ncs/zephyr/arch/arm/include -I/home/sebo/ncs/zephyr/include -I/home/sebo/ncs/zephyr/include/drivers -Izephyr/include/generated -I/home/sebo/ncs/zephyr/soc/arm/nordic_nrf/nrf52 -I/home/sebo/ncs/zephyr/soc/arm/nordic_nrf/include -I/home/sebo/ncs/zephyr/lib/libc/minimal/include -I/home/sebo/ncs/zephyr/ext/hal/cmsis/Include -I/home/sebo/ncs/zephyr/ext/hal/nordic/nrfx -I/home/sebo/ncs/zephyr/ext/hal/nordic/nrfx/drivers/include -I/home/sebo/ncs/zephyr/ext/hal/nordic/nrfx/hal -I/home/sebo/ncs/zephyr/ext/hal/nordic/nrfx/mdk -I/home/sebo/ncs/zephyr/ext/hal/nordic/. -I/home/sebo/ncs/zephyr/ext/debug/segger/rtt -I/home/sebo/ncs/nrf/include -I/home/sebo/ncs/zephyr/../cmock/vendor/unity/src -I/home/sebo/ncs/zephyr/../cmock/src -isystem /home/sebo/Downloads/gnu-mcu-eclipse/arm-none-eabi-gcc/8.2.1-1.4-20190214-0604/bin/../lib/gcc/arm-none-eabi/8.2.1/include -isystem /home/sebo/Downloads/gnu-mcu-eclipse/arm-none-eabi-gcc/8.2.1-1.4-20190214-0604/bin/../lib/gcc/arm-none-eabi/8.2.1/include-fixed -Os -nostdinc -g -Wall -Wformat -Wformat-security -Wno-format-zero-length -imacros /home/sebo/ncs/nrf/tests/unity/example_test/b/zephyr/include/generated/autoconf.h -ffreestanding -Wno-main -fno-common -mthumb -mcpu=cortex-m4 -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-strict-overflow -Wno-pointer-sign -Wno-unused-but-set-variable -fno-reorder-functions -fno-defer-pop -Werror=implicit-int -Wpointer-arith -ffunction-sections -fdata-sections -mabi=aapcs -march=armv7e-m -std=c99 -MD -MT CMakeFiles/app.dir/runner/runner/runner_example_test.c.obj -MF CMakeFiles/app.dir/runner/runner/runner_example_test.c.obj.d -o CMakeFiles/app.dir/runner/runner/runner_example_test.c.obj   -c runner/runner/runner_example_test.c
In file included from /home/sebo/ncs/zephyr/../cmock/vendor/unity/src/unity.h:16,
                 from runner/runner/runner_example_test.c:29:
/home/sebo/ncs/zephyr/../cmock/vendor/unity/src/unity_internals.h:15:10: fatal error: setjmp.h: No such file or directory
 #include <setjmp.h>
          ^~~~~~~~~~
compilation terminated.
[17/117] Building C object CMakeFiles/app.dir/mocks/foo/mock_foo.c.obj
FAILED: CMakeFiles/app.dir/mocks/foo/mock_foo.c.obj 
ccache /home/sebo/Downloads/gnu-mcu-eclipse/arm-none-eabi-gcc/8.2.1-1.4-20190214-0604/bin/arm-none-eabi-gcc -DBUILD_VERSION=v1.14.0-rc1-896-g578dac80f7c3 -DKERNEL -DNRF52840_XXAA -D_FORTIFY_SOURCE=2 -D__ZEPHYR__=1 -Imocks -I../src -I.././src/uut -I../. -I/home/sebo/ncs/zephyr/kernel/include -I/home/sebo/ncs/zephyr/arch/arm/include -I/home/sebo/ncs/zephyr/include -I/home/sebo/ncs/zephyr/include/drivers -Izephyr/include/generated -I/home/sebo/ncs/zephyr/soc/arm/nordic_nrf/nrf52 -I/home/sebo/ncs/zephyr/soc/arm/nordic_nrf/include -I/home/sebo/ncs/zephyr/lib/libc/minimal/include -I/home/sebo/ncs/zephyr/ext/hal/cmsis/Include -I/home/sebo/ncs/zephyr/ext/hal/nordic/nrfx -I/home/sebo/ncs/zephyr/ext/hal/nordic/nrfx/drivers/include -I/home/sebo/ncs/zephyr/ext/hal/nordic/nrfx/hal -I/home/sebo/ncs/zephyr/ext/hal/nordic/nrfx/mdk -I/home/sebo/ncs/zephyr/ext/hal/nordic/. -I/home/sebo/ncs/zephyr/ext/debug/segger/rtt -I/home/sebo/ncs/nrf/include -I/home/sebo/ncs/zephyr/../cmock/vendor/unity/src -I/home/sebo/ncs/zephyr/../cmock/src -isystem /home/sebo/Downloads/gnu-mcu-eclipse/arm-none-eabi-gcc/8.2.1-1.4-20190214-0604/bin/../lib/gcc/arm-none-eabi/8.2.1/include -isystem /home/sebo/Downloads/gnu-mcu-eclipse/arm-none-eabi-gcc/8.2.1-1.4-20190214-0604/bin/../lib/gcc/arm-none-eabi/8.2.1/include-fixed -Os -nostdinc -g -Wall -Wformat -Wformat-security -Wno-format-zero-length -imacros /home/sebo/ncs/nrf/tests/unity/example_test/b/zephyr/include/generated/autoconf.h -ffreestanding -Wno-main -fno-common -mthumb -mcpu=cortex-m4 -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-strict-overflow -Wno-pointer-sign -Wno-unused-but-set-variable -fno-reorder-functions -fno-defer-pop -Werror=implicit-int -Wpointer-arith -ffunction-sections -fdata-sections -mabi=aapcs -march=armv7e-m -std=c99 -MD -MT CMakeFiles/app.dir/mocks/foo/mock_foo.c.obj -MF CMakeFiles/app.dir/mocks/foo/mock_foo.c.obj.d -o CMakeFiles/app.dir/mocks/foo/mock_foo.c.obj   -c mocks/foo/mock_foo.c
mocks/foo/mock_foo.c:4:10: fatal error: setjmp.h: No such file or directory
 #include <setjmp.h>
          ^~~~~~~~~~
compilation terminated.
[18/117] Building C object CMakeFiles/app.dir/src/example_test.c.obj
FAILED: CMakeFiles/app.dir/src/example_test.c.obj 
ccache /home/sebo/Downloads/gnu-mcu-eclipse/arm-none-eabi-gcc/8.2.1-1.4-20190214-0604/bin/arm-none-eabi-gcc -DBUILD_VERSION=v1.14.0-rc1-896-g578dac80f7c3 -DKERNEL -DNRF52840_XXAA -D_FORTIFY_SOURCE=2 -D__ZEPHYR__=1 -Imocks -I../src -I.././src/uut -I../. -I/home/sebo/ncs/zephyr/kernel/include -I/home/sebo/ncs/zephyr/arch/arm/include -I/home/sebo/ncs/zephyr/include -I/home/sebo/ncs/zephyr/include/drivers -Izephyr/include/generated -I/home/sebo/ncs/zephyr/soc/arm/nordic_nrf/nrf52 -I/home/sebo/ncs/zephyr/soc/arm/nordic_nrf/include -I/home/sebo/ncs/zephyr/lib/libc/minimal/include -I/home/sebo/ncs/zephyr/ext/hal/cmsis/Include -I/home/sebo/ncs/zephyr/ext/hal/nordic/nrfx -I/home/sebo/ncs/zephyr/ext/hal/nordic/nrfx/drivers/include -I/home/sebo/ncs/zephyr/ext/hal/nordic/nrfx/hal -I/home/sebo/ncs/zephyr/ext/hal/nordic/nrfx/mdk -I/home/sebo/ncs/zephyr/ext/hal/nordic/. -I/home/sebo/ncs/zephyr/ext/debug/segger/rtt -I/home/sebo/ncs/nrf/include -I/home/sebo/ncs/zephyr/../cmock/vendor/unity/src -I/home/sebo/ncs/zephyr/../cmock/src -isystem /home/sebo/Downloads/gnu-mcu-eclipse/arm-none-eabi-gcc/8.2.1-1.4-20190214-0604/bin/../lib/gcc/arm-none-eabi/8.2.1/include -isystem /home/sebo/Downloads/gnu-mcu-eclipse/arm-none-eabi-gcc/8.2.1-1.4-20190214-0604/bin/../lib/gcc/arm-none-eabi/8.2.1/include-fixed -Os -nostdinc -g -Wall -Wformat -Wformat-security -Wno-format-zero-length -imacros /home/sebo/ncs/nrf/tests/unity/example_test/b/zephyr/include/generated/autoconf.h -ffreestanding -Wno-main -fno-common -mthumb -mcpu=cortex-m4 -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-strict-overflow -Wno-pointer-sign -Wno-unused-but-set-variable -fno-reorder-functions -fno-defer-pop -Werror=implicit-int -Wpointer-arith -ffunction-sections -fdata-sections -mabi=aapcs -march=armv7e-m -std=c99 -MD -MT CMakeFiles/app.dir/src/example_test.c.obj -MF CMakeFiles/app.dir/src/example_test.c.obj.d -o CMakeFiles/app.dir/src/example_test.c.obj   -c ../src/example_test.c
In file included from /home/sebo/ncs/zephyr/../cmock/vendor/unity/src/unity.h:16,
                 from ../src/example_test.c:6:
/home/sebo/ncs/zephyr/../cmock/vendor/unity/src/unity_internals.h:15:10: fatal error: setjmp.h: No such file or directory
 #include <setjmp.h>
          ^~~~~~~~~~
compilation terminated.
[19/117] Building C object zephyr/CMakeFiles/zephyr.dir/arch/common/sw_isr_common.c.obj
[20/117] Building C object zephyr/CMakeFiles/zephyr.dir/arch/common/isr_tables.c.obj
[21/117] Building C object zephyr/CMakeFiles/zephyr.dir/lib/os/thread_entry.c.obj
[22/117] Building C object zephyr/CMakeFiles/zephyr.dir/lib/os/work_q.c.obj
[23/117] Building C object zephyr/CMakeFiles/zephyr.dir/lib/os/fdtable.c.obj
[24/117] Building C object zephyr/CMakeFiles/zephyr.dir/lib/os/mempool.c.obj
[25/117] Building C object zephyr/CMakeFiles/zephyr.dir/lib/os/rb.c.obj
ninja: build stopped: subcommand failed.

```

nordic-krch replied:
currently it's native_posix. fixing `setjmp.h` error can come as incremental improvement. I'll add .rst

SebastianBoe replied:
Ah, k, got it working now :)

tejlmand replied:
Why was this resolved ?
I think what @SebastianBoe is asking for is a `sample.yaml` file in the folder:
`tests/unity/example_test`
with content similar to:
```
sample:
  name: example_test Sample
tests:
  test_build:
    build_only: true
    build_on_all: true
    platform_whitelist: native_posix
    tags: ci_build
```

nordic-krch replied:
added file https://github.com/NordicPlayground/fw-nrfconnect-nrf/blob/175a1bd3f6799e9c1cb2e872d19df36d4b97c709/tests/unity/example_test/sample.yaml

SebastianBoe commented on tests/unity/CMakeLists.txt:
```diff
@@ -0,0 +1,181 @@
+
+# Generate test runner file.
+function(test_runner_generate test_file_path)
+    get_property(RUBY_EXECUTABLE GLOBAL PROPERTY RUBY_EXECUTABLE)
```
indentation is two whitespaces

nordic-krch replied:
done

SebastianBoe commented on tests/unity/CMakeLists.txt:
```diff
@@ -0,0 +1,181 @@
+
+# Generate test runner file.
+function(test_runner_generate test_file_path)
+    get_property(RUBY_EXECUTABLE GLOBAL PROPERTY RUBY_EXECUTABLE)
```
find_program sets it's output parameter as a CACHE variable, which always has global scope, so we don't need to wrap the ruby executable in a property.

nordic-krch replied:
done

SebastianBoe commented:
I'm done with review, after the 'west' and build failure has been addressed this is OK.

carlescufi commented:
@SebastianBoe yep, we do need to clone recursively in west, but that would also mean fetching + updating submodules, so this requires 2 fixes:

1. `git clone --recurse` in `west init`
2. `git submodule update` in `west update`

@tejlmand can you take a look at this?

carlescufi commented on west.yml:
```diff
@@ -41,6 +43,10 @@ manifest:
       path: nrfxlib
       revision: d98b59bdff2b533f5afb7bdad229eaeb4a931882
+    - name: cmock
+      path: cmock
```
This should probably be somewhere deeper like `test/cmock`. Otherwise we will end up with a realy wide root folder. @SebastianBoe @tejlmand thoughts?

SebastianBoe replied:
No comment.

tejlmand replied:
Agree with @carlescufi .

Another reason why I would like this further down, is to make sure it the the most important part of NCS is located at toplevel, and then supporting tools, such as test tools are further down.
There may be more such tools in future.

Upstream Zephyr start to have `modules/` should we place it there instead of `test` ?

tejlmand commented on tests/unity/CMakeLists.txt:
```diff
@@ -0,0 +1,181 @@
+    execute_process(
+      COMMAND
+      ${PYTHON_EXECUTABLE}
+      ${ZEPHYR_BASE}/../nrf/scripts/unity/func_name_list.py
```
Why not use the `${NRF_DIR}` which is being defined in the root `<nrf>/CMakeLists.txt` file ?

SebastianBoe replied:
Because it is not set in this scope.

We could have NRF_DIR set as a CACHE variable to get around this maybe ...

nordic-krch replied:
done (https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/519/commits/44277184e3c09e27abc79329a35d83f0fd008d0b)

tejlmand commented on tests/unity/CMakeLists.txt:
```diff
@@ -0,0 +1,181 @@
+    execute_process(
+      COMMAND
+      ${PYTHON_EXECUTABLE}
+      ${ZEPHYR_BASE}/../nrf/scripts/unity/header_prepare.py
```
`${NRF_DIR}` ?

nordic-krch replied:
we had conversation about that with @SebastianBoe here https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/519#discussion_r267239960

I would like to if you tell me how to do it 😄

tejlmand replied:
Ahh, correct.
See you place `NRF_DIR` in the cache, which I like.
Does make code bit more flexible.

nordic-krch commented:
@tejlmand 
>Upstream Zephyr start to have modules/ should we place it there instead of test ?

maybe `ext` to indicate external dependency?

nordic-krch commented:
@tejlmand flake8 run on both scripts and fixed.

nordic-krch commented:
I've added user guide `Testing with Unity and CMock`, @ru-fu can you review it?

nordic-krch commented:
@carlescufi @tejlmand @SebastianBoe so where should i move repo
modules? ext? test? modules/test or ext/test? I can imagine that someone will add cmocka or cunit support and then ext/test folder may grow

carlescufi commented:
> @carlescufi @tejlmand @SebastianBoe so where should i move repo
> modules? ext? test? modules/test or ext/test? I can imagine that someone will add cmocka or cunit support and then ext/test folder may grow

My vote goes for `test/cmock`. If someone then adds another test framework that can live side-by-side: `test/cunit`

nordic-krch commented:
i've moved cmock to `test/cmock`.  For now i've added unity submodule explicitly to west.yml.

Pending:
- documentation review (@ru-fu)
- remove submodule from west.yml when @tejlmand adds submodule support to west (can happen after PR is merged)

tejlmand commented:
> so where should i move repo
> modules? ext? test? modules/test or ext/test? I can imagine that someone will add cmocka or cunit support and then ext/test folder may grow

I'm fine with all suggestions, with a slight preference for `test`

tejlmand commented on tests/unity/CMakeLists.txt:
```diff
@@ -0,0 +1,176 @@
+# - version with addtional __wrap_ prefix for all functions that
+#   is used to generate cmock
+function(cmock_headers_prepare in_header out_header wrap_header)
+  execute_process(
```
Are you sure you want to use `execute_process` here ?
This means that the header file provided is ONLY processed at configure time.

When looking here: https://github.com/NordicPlayground/fw-nrfconnect-nrf/blob/93ac00157f2027b8d6516be824841e1f0bbb45b9/tests/unity/example_test/CMakeLists.txt#L16

you are providing the header to use.
Now, if a user modifies `src/foo/foo.h` and the rebuilds in this example, then `header_prepare.py` WILL NOT be executed again, and the  mocks are still based on the old header file.
`header_prepare.py` will not be executed until CMake is executed again (either manually or because of other dependencies).

If you want `header_prepare.py` to execute everytime `foo.h` is modified, then you should probably look at: 
`add_custom_command`, https://cmake.org/cmake/help/latest/command/add_custom_command.html

`add_custom_command` will also allow you to set up more dependencies for the output from `header_prepare.py` to ensure those depending on that output are updated as well.

nordic-krch replied:
I've added 'add_custom_command` to generation of test runner. This is in fact very useful because adding a new test case will not require any action. Runner will be regenerated when test file is modified.

Regarding using the same approach for generation of cmocks. For now i didn't add it. It seems to me more complex as it's not only replacing `execute_process` with `add_custom_command` because cmock generation process consists of few steps, some are calling cmake functions and some are calling external command (python). It is also very unlikely that this will be needed because if we are working on writing a test for given module it's rather unlikely that we will be changing mocked API at the same time. As I would like to get that in, i considered it as not must have for now. This can be added in future if users see the need. I guess they won't be that affected.

tejlmand replied:
> It is also very unlikely that this will be needed because if we are working on writing a test for given module it's rather unlikely that we will be changing mocked API at the same time. 

Such assumptions of how user will or will not behave is like placing a trap for those working in a different way, and just waiting for them to fall into it.
Those user won't have the knowledge about this, that we have.

What if I'm working on the module that is being mocked else-where, and I as part of my changes run `build/run all unittests` (not possible in current design AFAIK), then the unit tests being built might suddenly cause build failure, due to mismatch between old mocks and updated header.
How is a user expected to be aware that `cmake` must be executed in order to get the code to compile ?

nordic-krch replied:
@tejlmand i'm not saying that it's not possible. What i wanted to say is that currently i don't know how to do that. If you can help me with that i'm happy if it's added. If not then taking into account priority of this feature i would prefer to get PR merged and improve incrementally.

nordic-krch replied:
@tejlmand ping. Do you agree with doing that as separate PR?

tejlmand replied:
after taking a closer look, I can see that the parts with execute process can easily be adjusted, however it also became apparent that output from one other the functions returns the functions to wrap and that information is used as argument to the linker command.
As the linker command is constructed at configure time, and does not appear to support providing a file containinig a list of functions to `ld`, then i'm fine with postponing this improvement till a later time.

Could you create a followup issue, so we don't forget ?
Priority: Low

nordic-krch replied:
Issue `NCSDK-1679` created

tejlmand commented on scripts/unity/header_prepare.py:
```diff
@@ -0,0 +1,58 @@
+
+    # change static inline functions to normal function declaration
+    static_inline_pattern = re.compile(
+        r'((__deprecated )?)((static inline )|(static ALWAYS_INLINE )|'
```
This regex is vulnerable to unintended additional spaces. as example if programmer by accident writes:
```
static  inline
``` 
with 2 spaces instead of one, it will fail.
And for the programmer the reason why `header_prepare.py` generates wrong (missing) output will be hard to track down, as a space too much in a header file might not be the first place to consider.

Second thing: Low priority.
regex is also vulnerable to line breaks in function declaration.
Although it's not used in NCS, I have seen that principle in code elsewhere, so it's up to you to decide if that should be handled as well.

nordic-krch replied:
replaced spaces with `\s+`.

tejlmand commented on tests/unity/example_test/Kconfig:
```diff
@@ -0,0 +1,17 @@
+
+mainmenu "Example unity test"
+
+menu "UUT module configuration"
```
I'm not sure a user opening `menuconfig` for this example knows what UUT means.
At least I don't !

nordic-krch replied:
Fixed, replaced with full name (unit under test)

tejlmand commented on tests/unity/CMakeLists.txt:
```diff
@@ -0,0 +1,176 @@
+  cmock_linker_wrap_trick(${mod_header_path})
+
+  target_include_directories(app PRIVATE ${CMOCK_PRODUCTS_DIR})
+  file(GLOB app_sources ${dst_path}/*.c)
```
Globbing should be used with care.
The files under `${dst_path}` are all generated by the build system, correct ?
In that case I will assume that we ought to know somewhere else, exactly what files has been generated, and can use that information.

If that is not the case, then this code might be ok, as is.

nordic-krch replied:
removed use of glob. Adding files explicitly.

nordic-krch commented:
@ru-fu please not that i've just updated documentation and removed last paragraph.

tejlmand commented on scripts/unity/func_name_list.py:
```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/python
```
Zephyr uses python3, so I guess this is safer:
```suggestion
#!/usr/bin/env python3
```
as some systems may have python pointing to `python 2.x` although I do know when called through CMake the specific python interpreter is used.

tejlmand commented on scripts/unity/header_prepare.py:
```diff
@@ -0,0 +1,58 @@
+#!/usr/bin/python
```
Zephyr uses python3, so I guess this is safer:
```suggestion
#!/usr/bin/env python3
```
as some systems may have python pointing to `python 2.x` although I do know when called through CMake the specific python interpreter is used.

tejlmand commented on tests/Kconfig:
```diff
@@ -0,0 +1,7 @@
+#
+# Copyright (c) 2018 Nordic Semiconductor
```
Is the year correct ?

tejlmand replied:
there are a couple of more places with year 2018.

nordic-krch commented:
@ru-fu thanks for review and changes. They look ok, i've squashed your commit.

nordic-krch commented:
@tejlmand follow-up issue created. `sample.yaml` created. Can you re-check?

nordic-krch commented:
@carlescufi i think we can merge it. Cmake experts are done, as well as the documentation.

carlescufi commented:
@nordic-krch very well, can you rebase on top of master?

nordic-krch commented:
@carlescufi rebased.

carlescufi commented:
@nordic-krch still shows conflicted, probably someone has merged stuff in the meantime.

nordic-krch commented:
@carlescufi can you merge now?

metc commented on tests/unity/example_test/Kconfig:
```diff
@@ -0,0 +1,17 @@
+# Kconfig - Config options for logger sample app
```
Nit: the comment is a bit off.
